File: 프로그래머스/2/12978. 배달/배달.py
from collections import defaultdict
import heapq


def solution(N, road, K):
    graph = defaultdict(list)
    for a, b, w in road:
        graph[a].append((w, b))
        graph[b].append((w, a))
    
    dist = dijkstra(1, N, graph)

    answer = 1
    for i in range(2, N+1):
        if dist[i] <= K:
            answer += 1

    return answer


def dijkstra(start, n, graph):
    dist = {i: float('inf') for i in range(1, n+1)}
    dist[start] = 0
    hq = []
    heapq.heappush(hq, (0, start))
    while hq:
        d1, node = heapq.heappop(hq)
        if d1 > dist[node]:
            continue
        for d2, next_node in graph[node]:
            d = d1 + d2
            if dist[next_node] > d:
                dist[next_node] = d
                heapq.heappush(hq, (d, next_node))
    
    return dist

File: 프로그래머스/2/12978. 배달/배달.java
import java.util.*;

class Solution {
    static List<List<Node>> graph;
    static int INF = Integer.MAX_VALUE;
    
    class Node implements Comparable<Node> {
        int n;
        int w;
        
        Node(int n, int w) {
            this.n = n;
            this.w = w;
        }
        
        @Override
        public int compareTo(Node node) {
            return this.w - node.w;
        }
    }
    
    public int solution(int N, int[][] road, int K) {
        graph = new ArrayList<>();
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int[] r : road) {
            int a = r[0], b = r[1], w = r[2];
            graph.get(a).add(new Node(b, w));
            graph.get(b).add(new Node(a, w));
        }
        
        int[] dist = dijkstra(1, N);
        int answer = 0;
        for (int i = 0; i <= N; i++) {
            if (dist[i] <= K) answer++;
        }

        return answer;
    }
    
    private int[] dijkstra(int start, int n) {
        int[] dist = new int[n + 1];
        Arrays.fill(dist, INF);
        dist[start] = 0;
        
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(start, 0));
        while (!pq.isEmpty()) {
            Node node = pq.poll();
            if (node.w > dist[node.n]) continue;
            for (Node nextNode : graph.get(node.n)) {
                int cost = node.w + nextNode.w;
                if (dist[nextNode.n] > cost) {
                    dist[nextNode.n] = cost;
                    pq.offer(new Node(nextNode.n, cost));
                }
            }
        }
        return dist;
    }
}